engine/src/main/java/org/pentaho/di/core/util/JavaScriptUtils.java

```diff
@@ -0,0 +1,234 @@
+        || classType.equalsIgnoreCase( "java.util.Date" ) ) {
+        // Is it a java Date() class ?
+        try {
+          Date dat = (Date) Context.jsToJava( value, java.util.Date.class );
+          dbl = dat.getTime();
+        } catch ( Exception e ) {
+          // Is it a Value?
+          //
+          return convertValueToDate( value );
+        }
+      } else if ( classType.equalsIgnoreCase( "java.lang.Double" ) ) {
+        dbl = (Double) value;
+      } else {
+        String string = (String) Context.jsToJava( value, String.class );
+        dbl = Double.parseDouble( string );
+      }
+      long lng = Math.round( dbl );
+      return new Date( lng );
+    }
+    return null;
+  }
+
+  public static BigDecimal jsToBigNumber( Object value, String classType ) {
+    if ( classType.equalsIgnoreCase( JS_UNDEFINED ) ) {
+      return null;
+    } else if ( classType.equalsIgnoreCase( JS_NATIVE_NUM ) ) {
+      Number nb = Context.toNumber( value );
+      return BigDecimal.valueOf( nb.doubleValue() );
+    } else if ( classType.equalsIgnoreCase( JS_NATIVE_JAVA_OBJ ) ) {
+      // Is it a BigDecimal class ?
+      return convertNativeJavaToBigDecimal( value );
+    } else if ( classType.equalsIgnoreCase( "java.lang.Byte" ) ) {
+      return BigDecimal.valueOf( ( (Byte) value ).longValue() );
+    } else if ( classType.equalsIgnoreCase( "java.lang.Short" ) ) {
+      return BigDecimal.valueOf( ( (Short) value ).longValue() );
+    } else if ( classType.equalsIgnoreCase( "java.lang.Integer" ) ) {
+      return BigDecimal.valueOf( ( (Integer) value ).longValue() );
+    } else if ( classType.equalsIgnoreCase( "java.lang.Long" ) ) {
+      return BigDecimal.valueOf( ( (Long) value ).longValue() );
+    } else if ( classType.equalsIgnoreCase( "java.lang.Double" ) ) {
+      return BigDecimal.valueOf( ( (Double) value ).doubleValue() );
+    } else if ( classType.equalsIgnoreCase( "java.lang.String" ) ) {
+      return BigDecimal.valueOf( ( new Long( (String) value ) ).longValue() );
+    } else {
+      throw new UnsupportedOperationException( "JavaScript conversion to BigNumber not implemented for " + classType );
+    }
+  }
+
+  private static BigDecimal convertNativeJavaToBigDecimal( Object value ) {
+    try {
+      return (BigDecimal) Context.jsToJava( value, BigDecimal.class );
+    } catch ( Exception e ) {
+      try {
+        Value v = (Value) Context.jsToJava( value, Value.class );
+        if ( !v.isNull() ) {
+          return v.getBigNumber();
+        } else {
+          return null;
+        }
+      } catch ( Exception e2 ) {
+        String string = (String) Context.jsToJava( value, String.class );
+        return new BigDecimal( string );
+      }
+    }
+  }
+
+  private static Date convertValueToDate( Object value ) throws KettleValueException {
+    try {
+      Value v = (Value) Context.jsToJava( value, Value.class );
+      return v.getDate();
+    } catch ( Exception e2 ) {

```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 229.